dgraph/cmd/graphql/schema/gqlschema.go

```diff
@@ -275,11 +363,196 @@ func addPatchType(schema *ast.Schema, defn *ast.Definition) {
-		Kind:   ast.InputObject,
-		Name:   defn.Name + "Filter",
-		Fields: getFilterField(),
+	if hasSearchables(defn) {
+		filterName := defn.Name + "Filter"
+		filter := &ast.Definition{
+			Kind: ast.InputObject,
+			Name: filterName,
+		}
+
+		for _, fld := range defn.Fields {
+			if searchable := getSearchable(fld); searchable != "" {
+				filterTypeName := builtInFilters[searchable]
+				if schema.Types[fld.Type.Name()].Kind == ast.Enum {
+					// If the field is an enum type, we don't generate a filter type.
+					// Instead we allow to write `fieldName: enumValue` in the filter.
+					// So, for example : `filter: { postType: Answer }`
+					// rather than : `filter: { postType: { eq: Answer } }`
+					//
+					// Booleans are the same, allowing:
+					// `filter: { isPublished: true }
+					// but that case is aready handled by builtInFilters
+					filterTypeName = fld.Type.Name()
+				}
+
+				filter.Fields = append(filter.Fields,
+					&ast.FieldDefinition{
+						Name: fld.Name,
+						Type: &ast.Type{
+							NamedType: filterTypeName,
+						},
+					})
+			}
+		}
+
+		filter.Fields = append(filter.Fields,
+			&ast.FieldDefinition{Name: "and", Type: &ast.Type{NamedType: filterName}},
+			&ast.FieldDefinition{Name: "or", Type: &ast.Type{NamedType: filterName}},
+			&ast.FieldDefinition{Name: "not", Type: &ast.Type{NamedType: filterName}},
+		)
+
+		schema.Types[filterName] = filter
+	}
+}
+
+func hasSearchables(defn *ast.Definition) bool {
+	return fieldAny(defn.Fields,
+		func(fld *ast.FieldDefinition) bool { return getSearchable(fld) != "" })
+}
+
+func hasOrderables(defn *ast.Definition) bool {
+	return fieldAny(defn.Fields,
+		func(fld *ast.FieldDefinition) bool { return orderable[fld.Type.Name()] })
+}
+
+// fieldAny returns true if any field in fields satisfies pred
+func fieldAny(fields ast.FieldList, pred func(*ast.FieldDefinition) bool) bool {
+	for _, fld := range fields {
+		if pred(fld) {
+			return true
+		}
+	}
+	return false
+}
+
+// getSearchable returns the name of the searchable applied to fld, or ""
+// if fld doesn't have a searchable directive.
+func getSearchable(fld *ast.FieldDefinition) string {
+	search := fld.Directives.ForName(searchableDirective)
+	if search == nil {
+		return ""
+	}
+	if len(search.Arguments) == 0 {
+		if searchable, ok := defaultSearchables[fld.Type.Name()]; ok {
+			return searchable
+		}
+		// it's an enum - always has exact index
+		return "exact"
+	}
+	return search.Arguments.ForName(searchableArg).Value.Raw
+}
+
+// addTypeOrderable adds an input type that allows ordering in query.
+// Two things are added: an enum with the names of all the orderable fields,
+// that's called TOrderable; and an input type that allows saying order asc
+// or desc, that's called TOrder.  TOrder's fileds are TOrderable's.  So you
+// might get:
+// enum PostOrderable { datePublished, numLikes, ... }, and
+// input PostOrder { asc : PostOrderable, desc: PostOrderable ...}
```

Comment:
Could you also mention `then` here and talk about how how it allows sorting by multiple predicates?